ports/zephyr/zephyr_device.c

```diff
@@ -27,23 +27,198 @@
 #include "zephyr_device.h"
 #include "py/runtime.h"
 
+#if defined(CONFIG_MICROPY_DYNAMIC_PINCTRL)
+
+#include <zephyr/drivers/pinctrl.h>
+extern pinctrl_soc_pin_t zephyr_pinctrl_get_pinctrl(const mp_obj_t obj);
+
+#endif
+
+/* Circumvent limitation of only having devices that are initialized */
+typedef struct mp_zephyr_device_data_t_ {
+    const struct device *dev;
+    char *name;
+    const char **labels;
+    size_t label_cnt;
+    #if defined(CONFIG_MICROPY_DYNAMIC_PINCTRL)
+    struct pinctrl_dev_config *pinctrl_cfg;
+    #endif
+} mp_zephyr_device_data_t;
+
+#define DT_NODELABEL_STRING_ARRAY_DECLARE(node) static const char *_##node##_labels[] = \
+    DT_NODELABEL_STRING_ARRAY(node);
+
+DT_FOREACH_STATUS_OKAY_NODE(DT_NODELABEL_STRING_ARRAY_DECLARE)
+
+#define DT_NODELABEL_STRING_ARRAY_NAME(node) _##node##_labels
+
+#define PINCTRL_DT_DEV_CONFIG_ENTRY(node) COND_CODE_1(DT_CAT(node, _PINCTRL_IDX_0_EXISTS), \
+    (PINCTRL_DT_DEV_CONFIG_GET(node)), (NULL))
+
+#if defined(CONFIG_MICROPY_DYNAMIC_PINCTRL)
+
+#define PINCTRL_DT_DATA_DECLARE_WEAKENED(node) \
+    extern Z_PINCTRL_DEV_CONFIG_CONST struct pinctrl_dev_config __attribute__((weak)) \
+    Z_PINCTRL_DEV_CONFIG_NAME(node);
+
+DT_FOREACH_STATUS_OKAY_NODE(PINCTRL_DT_DATA_DECLARE_WEAKENED)
+
+#define DEVICE_DATA_ENTRY(node) {                        \
+        .dev = DEVICE_DT_GET(node),                      \
+        .name = DEVICE_DT_NAME(node),                    \
+        .labels = DT_NODELABEL_STRING_ARRAY_NAME(node),  \
+        .label_cnt = DT_CAT(node, _NODELABEL_NUM),       \
+        .pinctrl_cfg = PINCTRL_DT_DEV_CONFIG_ENTRY(node) \
+},
+
+#else
+
+#define DEVICE_DATA_ENTRY(node) {                       \
+        .dev = DEVICE_DT_GET(node),                     \
+        .name = DEVICE_DT_NAME(node),                   \
+        .labels = DT_NODELABEL_STRING_ARRAY_NAME(node), \
+        .label_cnt = DT_CAT(node, _NODELABEL_NUM),      \
+},
+
+#endif
+
+#define WEAKEN_DEVICE(node) extern const struct device __attribute__((weak)) (DEVICE_DT_NAME_GET(node));
+
+DT_FOREACH_STATUS_OKAY_NODE(WEAKEN_DEVICE)
```

Comment:
I didn't know it was possible to 'weaken' an external symbol!
It seems like this is not used anywhere anymore though?
Or is this required to get `DEVICE_DATA_ENTRY()` to work?

Reply:
I was surprised this worked at all yea.

Yes, it is required, otherwise the table gets undefined references to devices, and I havent figured a way to check that at build time.

Reply:
Okay so that method definitely doesnt work.
It's also not possible to filter by _PINCTRL_IDX_0_EXISTS for devices without pinctrl, which is a problem for devices without a default pinctlr so that exposed that issue.

Reply:
It seems to update pinctrl the device MUST have a pinctrl_cfg struct allocated at build time...

Reply:
> It seems to update pinctrl the device MUST have a pinctrl_cfg struct allocated at build time... That is because it's stored in a device driver defined struct and **gotten from dts** at build time.

wow so `_PINCTRL_IDX_0_EXISTS` works after all?

> This bit is actually intended to list all devices.

That was useful for getting nodelabels, but maybe it is not needed for getting pinctrl to work.